docs/adr/adr-005-plugins.md

```diff
@@ -0,0 +1,273 @@
+# ADR 005: Plugins
+
+## Changelog
+
+- 2022-02-06: init commit
+
+## Context
+
+The modular design of celestia-node allows for the functionality it provides to be extended. The main blocker for creating custom light client applications is having a mechanism to access the internals of a celestia-node. Here we are suggesting that there be some interface that installs custom services into celestia-node, along with making the celestia-node cli more portable, so that light client developers can create their own binaries with the custom services that they create.
```

Comment:
I guess my only question for this design is what the difference is between us adding in this plugin functionality vs us creating a robust enough `LightNode` or `FullNode` RPC interface that would enable the same kind of custom client? I don't really see the difference between Plugins and providing a super robust `LightNode` RPC interface.

Reply:
as mentioned [here](https://github.com/celestiaorg/celestia-node/pull/414#issuecomment-1055871523), I do think that in the short-medium term, we should simply use a simplified rpc and no plugins. 

However, this is only possible because recently we have reduced the scope of how agnostic optimint needs to be, and are willing to copy paste code from celestia-core instead of importing. If we still wanted those features, with a robust RPC we would have to deal with all the UX terribleness of having one extra binary to run, handling yet another cli, redundant testing, and redundant basic program structure. With plugins we don't.

Plugins would also enable entirely new node types, like custom DAS, storage, rpc, and even custom blockchains that have a one way bridge from celestia. These custom node types might be possible over rpc, but they would suffer the same ux issues as the dalc does, and might even be reliant on us adding specific rpc methods that we otherwise would not wish to expose.

again, I think designing for those things so early is a mistake, and we should go with a simplified dalc-esque rpc instead, but I think having in process access to celestia-node internals would prove invaluable to those custom node types.